src/context/internal_nostd.rs

```diff
@@ -0,0 +1,159 @@
+            f(&*secp)
+        },
+        rerandomize_seed,
+    )
+}
+
+/// Borrows the global context as a raw pointer and do some operation on it.
+///
+/// If `randomize_seed` is provided, it is used to rerandomize the context after the
+/// operation is complete. If it is not provided, randomization is skipped.
+///
+/// Only a bit or two per signing operation is needed; if you have any entropy at all,
+/// you should provide it, even if you can't provide 32 random bytes.
+pub fn with_raw_global_context<T, F: FnOnce(NonNull<ffi::Context>) -> T>(
+    f: F,
+    rerandomize_seed: Option<&[u8; 32]>,
+) -> T {
+    // Our function may be expensive, so before calling it, we copy the global
+    // context into this local buffer on the stack. Then we can release it,
+    // allowing other callers to use it simultaneously.
+    let mut ctx = SelfContainedContext::new_uninitialized();
+    let mut have_global_ctx = false;
+    if let Some(mut guard) = SECP256K1.try_lock() {
+        let global_ctx = &mut *guard;
+        ctx.clone_into(global_ctx);
+        have_global_ctx = true;
+        // (the lock is now dropped)
+    }
+
+    // Obtain a raw pointer to the context, creating one if it has not been already,
+    // and call the function.
+    let ctx_ptr = ctx.raw_ctx();
+    let ret = f(ctx_ptr);
+
+    // ...then rerandomize the local copy, and try to replace the global one
+    // with this. Note that even if we got the lock above, we may fail to get
+    // it now; in that case, we don't rerandomize and leave the contexct in
+    // the state that we found it in.
+    //
+    // We do this, rather than holding the lock continuously through the call
+    // to `f`, to minimize the likelihood of contention. If we fail to randomize,
+    // that really isn't a big deal since this is a "defense in depth" measure
+    // whose value is likely to obtain even if it only succeeds a small fraction
+    // of the time.
+    //
+    // Contention, meanwhile, will lead to users using a stack-local copy of
+    // the context rather than the global one, which aside from being inefficient,
+    // means that the context they use won't be rerandomized at all. So there
+    // isn't even any benefit.
+    if have_global_ctx {
+        if let Some(seed) = rerandomize_seed {
+            // SAFETY: just a FFI call
+            unsafe {
+                assert_eq!(ffi::secp256k1_context_randomize(ctx_ptr, seed.as_ptr()), 1);
+            }
+            if let Some(ref mut guard) = SECP256K1.try_lock() {
+                guard.clone_into(&mut ctx);
+            }
+        }
+    }
+    ret
+}
+
+/// Rerandomize the global context, using the given data as a seed.
+///
+/// The provided data will be mixed with the entropy from previous calls in a timing
+/// analysis resistant way. It is safe to directly pass secret data to this function.
+pub fn rerandomize_global_context(seed: &[u8; 32]) { with_raw_global_context(|_| {}, Some(seed)) }
```

Comment:
Maybe add a simple test for no-std that verifies basic functionality ?

```suggestion
pub fn rerandomize_global_context(seed: &[u8; 32]) { with_raw_global_context(|_| {}, Some(seed)) }

#[cfg(all(test, not(feature = "std")))]
mod nostd_tests {
    use super::*;
    use crate::{Message, SecretKey, Secp256k1};

    #[test]
    fn basic_global_context_usage() {
        let msg = Message::from_digest(crate::test_random_32_bytes());
        let sk = SecretKey::test_random();
        
        let sig = crate::with_global_context(
            |secp: &Secp256k1<crate::SignOnly>| secp.sign_ecdsa(msg, &sk),
            Some(&sk.secret_bytes()),
        );
        
        let pk = crate::with_global_context(
            |secp: &Secp256k1<crate::SignOnly>| crate::key::PublicKey::from_secret_key(secp, &sk),
            None,
        );
        
        let result = crate::with_global_context(
            |secp: &Secp256k1<crate::VerifyOnly>| secp.verify_ecdsa(&sig, msg, &pk),
            None,
        );
        
        assert!(result.is_ok());
    }
}
```